packages/react/src/evaluation/use-feature-flag.ts

```diff
@@ -52,6 +53,57 @@ enum SuspendState {
   Error,
 }
 
+// just used for casting, etc, but don't use for return values because the name isn't as clear
+type ConstrainedFlagQuery<T> = FlagQuery<
+  T extends boolean
+    ? boolean
+    : T extends number
+      ? number
+      : T extends string
+        ? string
+        : T extends JsonValue
+          ? T
+          : JsonValue
+>;
```

Comment:
This type is a bit wild-looking, we need it. We have to use the conditional, because otherwise `useFlag('key', false)` would return `false`, not `boolean` (too constrained).

Reply:
We need [NoInfer](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#the-noinfer-utility-type) but that would require TypeScript 5.4. Maybe we could use a polyfill for now.

Reply:
Maybe, but if you look closely, we actually DO want to infer that last condition, so that specific object generic args work nicely:

```
const { value: objVal }: FlagQuery<{greeting: string}> = useFlag('obj-flag', { greeting: 'hi' }); // we don't want a `JsonValue` here, we want a `T` (in this case a `greeting: string}`.
```